### PR TITLE
Add a method to the arena to promote objects to it if they implement LocalDrop

### PR DESCRIPTION
**Pull request by Zoxc**
A cleaned up part from https://github.com/rust-lang/rust/pull/56382.

Uses PROC MACROS! cc @eddyb 

r? @michaelwoerister

### CODE REVIEW

File: src/librustc_data_structures/local_drop.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::collections::{HashMap, HashSet};
+use std::hash::BuildHasherDefault;
+
+/// A type which does not look at references in its destructor
+/// but only owned data
+pub unsafe trait LocalDrop {
```
Comment (eddyb) [inline src/librustc_data_structures/local_drop.rs]: cc @pnkfelix @nikomatsakis

**Comment by michaelwoerister** [comment]
@Zoxc, can you give some explanation and reasoning of the concepts any changes here? What is `LocalDrop` and why do we need it? What is "promotion" in this context?

**Comment by Zoxc** [comment]
Promotion would be allocating types with destructors on the dropless arena by deferring execution of their destructors to when the arena is dropped. We can only allow that for types which do not access other promoted objects during destruction, since those promoted object may have been destroyed earlier. Such allowed types implement `LocalDrop`. Types without a `Drop` `impl` trivially have that property and this allows you to `#[derive]` that fact.

**Comment by eddyb** [comment]
The point of the dropless arena is that, well, it has no destructors, simplifying a few things. You can add separate regular typed arenas for things you can't make dropless.

**Comment by michaelwoerister** [comment]
To be honest, that sounds a bit too magical for my taste. cc @rust-lang/compiler

**Comment by pnkfelix** [comment]
@eddyb wrote:

> The point of the dropless arena is that, well, it has no destructors, simplifying a few things. You can add separate regular typed arenas for things you can't make dropless.

I want to clarify the point eddyb is making here.

@eddyb, are you asking for @Zoxc to revise this PR to get rid of `LocalDrop`, and instead just distinguish between a dropless arena which would presumably only accept ... things that are `Copy`?

Or is your point just to explain what the purpose of `LocalDrop` is, and when you said "no destructors", you meant something more nuanced, like (perhaps) "no destructors with non-trivial side-effects -- i.e. no destructors that touch anything except for data they own that is also `LocalDrop`" ?

**Comment by eddyb** [comment]
@pnkfelix The former, except it's not a `Copy` bound but an `assert!(!needs_drop::<T>())`.

**Pull request closed without merging**